### ISSUE TITLE
[CURATOR-472] Test-only code started to be called in prod

### ISSUE THREAD

**Issue by jira-importer** [apache/curator#992]
<p>After commit <tt>fe2c7c4cd606c0cf4bc4fab15deedc0f4c33ea0e</tt> the method</p>

<p> </p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">
<span class="code-keyword">protected</span> <span class="code-object">boolean</span> testableWaitForShutdown(<span class="code-object">int</span> wait)
    <span class="code-keyword">throws</span> InterruptedException
</pre>
</div></div>
<p>with the following comment</p>

<p> </p>

<p> </p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">
* Wait up to wait milliseconds <span class="code-keyword">for</span> the underlying threads to shutdown.
* THIS METHOD IS EXPECTED TO BE USED FOR TESTING ONLY!!!
</pre>
</div></div>
<p>started to be called from real method <tt>ZooKeeper.close()</tt></p>

<p>It prevents tests to finish.</p>

<p> </p>

---
<details><summary><i>Originally reported by <a title="mskonovalov" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mskonovalov">mskonovalov</a>, imported from: <a href="https://issues.apache.org/jira/browse/CURATOR-472" target="_blank">Test-only code started to be called in prod</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a title="Jordan Zimmerman" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=randgalt">randgalt</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Critical
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2018-12-03T01:18:14+00:00
<li><b>imported</b>: 2025-01-21
</ul></i>
</details>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cammckenzie">cammckenzie</a>:</i>
<p>This is interesting, as the actual call to close() on the Zookeeper handle (passing in the time to wait in MS) is not documented as a 'testing only' interface.</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">
/**
* Close <span class="code-keyword">this</span> client object as the {@link #close() } method.
* This method will wait <span class="code-keyword">for</span> internal resources to be released.
*
* @param waitForShutdownTimeoutMs timeout (in milliseconds) to wait <span class="code-keyword">for</span> resources to be released.
* Use zero or a negative value to skip the wait
* @<span class="code-keyword">throws</span> InterruptedException
* @<span class="code-keyword">return</span> <span class="code-keyword">true</span> <span class="code-keyword">if</span> waitForShutdownTimeout is greater than zero and all of the resources have been released
*
* @since 3.5.4
*/
<span class="code-keyword">public</span> <span class="code-object">boolean</span> close(<span class="code-object">int</span> waitForShutdownTimeoutMs) <span class="code-keyword">throws</span> InterruptedException {</pre>
</div></div>
<p>But, internally, this method calls the testableWaitForShutdown method which is marked as 'for testing only'.</p>

<p>I also notice that it is stopping tests from completing, but I haven't established why this is occurring yet.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mskonovalov">mskonovalov</a>:</i>
<p>It seems to be blocked on joining the thread </p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">
sendThread
</pre>
</div></div>
<p> </p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cammckenzie">cammckenzie</a>:</i>
<p>Thanks for the update. So, my initial thoughts are that this is a Zookeeper issue. Either, it's a documentation issue in that the public API they're exposing makes no mention of the close(int) method not being safe for clients to call. Or, there's an actual issue with the shutdown process.</p>

<p>It doesn't appear the Curator is doing anything strange with its interaction with ZK. Thoughts?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mskonovalov">mskonovalov</a>:</i>
<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cammckenzie" class="user-hover" rel="cammckenzie">Cam McKenzie</a> are you able to reproduce the issue?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cammckenzie">cammckenzie</a>:</i>
<p>Yeah, I believe so, the tests are hanging for me on the master branch, but I have not had a chance to investigate beyond that yet.</p>

<p> </p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mskonovalov">mskonovalov</a>:</i>
<p><tt>The sendThread is in SLEEPING state when we are trying to join it</tt></p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cammckenzie">cammckenzie</a>:</i>
<p>Thanks, I would have thought that when the close() is called on ZK that closing the underlying connection would cause the sendThread to wake up, but I haven't looked at the ZK code, so can't confirm what they're sleeping on. Maybe there's a race condition on shutdown?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mskonovalov">mskonovalov</a>:</i>
<p>Here is dome thread dumps as well</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">
<span class="code-quote">"main@1"</span> prio=5 tid=0x1 nid=NA waiting <span class="code-keyword">for</span> monitor entry
java.lang.<span class="code-object">Thread</span>.State: BLOCKED
waiting <span class="code-keyword">for</span> main-EventThread@1594 to release lock on <0x716> (a org.apache.curator.ConnectionState)
at org.apache.curator.ConnectionState.checkTimeouts(ConnectionState.java:209)
at org.apache.curator.ConnectionState.getZooKeeper(ConnectionState.java:96)
at org.apache.curator.CuratorZookeeperClient.getZooKeeper(CuratorZookeeperClient.java:164)
at org.apache.curator.framework.imps.CuratorFrameworkImpl.getZooKeeper(CuratorFrameworkImpl.java:597)
at org.apache.curator.framework.imps.ExistsBuilderImpl$3.call(ExistsBuilderImpl.java:268)
at org.apache.curator.framework.imps.ExistsBuilderImpl$3.call(ExistsBuilderImpl.java:257)
at org.apache.curator.connection.StandardConnectionHandlingPolicy.callWithRetry(StandardConnectionHandlingPolicy.java:64)
at org.apache.curator.RetryLoop.callWithRetry(RetryLoop.java:100)
at org.apache.curator.framework.imps.ExistsBuilderImpl.pathInForegroundStandard(ExistsBuilderImpl.java:254)
at org.apache.curator.framework.imps.ExistsBuilderImpl.pathInForeground(ExistsBuilderImpl.java:247)
at org.apache.curator.framework.imps.ExistsBuilderImpl.forPath(ExistsBuilderImpl.java:206)
at org.apache.curator.framework.imps.ExistsBuilderImpl.forPath(ExistsBuilderImpl.java:35)
at org.apache.curator.framework.imps.Test.testFastConnect(Test.java:21)
at sun.reflect.NativeMethodAccessorImpl.invoke0(NativeMethodAccessorImpl.java:-1)
at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
at java.lang.reflect.Method.invoke(Method.java:498)
at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
at org.testng.internal.Invoker.invokeMethod(Invoker.java:645)
at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
at org.testng.TestRunner.privateRun(TestRunner.java:756)
at org.testng.TestRunner.run(TestRunner.java:610)
at org.testng.SuiteRunner.runTest(SuiteRunner.java:387)
at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:382)
at org.testng.SuiteRunner.privateRun(SuiteRunner.java:340)
at org.testng.SuiteRunner.run(SuiteRunner.java:289)
at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
at org.testng.TestNG.runSuitesSequentially(TestNG.java:1293)
at org.testng.TestNG.runSuitesLocally(TestNG.java:1218)
at org.testng.TestNG.runSuites(TestNG.java:1133)
at org.testng.TestNG.run(TestNG.java:1104)
at org.testng.IDEARemoteTestNG.run(IDEARemoteTestNG.java:72)
at org.testng.RemoteTestNGStarter.main(RemoteTestNGStarter.java:123)

</pre>
</div></div>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">
 <span class="code-quote">"main-SendThread(192.0.2.0:2181)@1593"</span> daemon prio=5 tid=0xf nid=NA sleeping java.lang.<span class="code-object">Thread</span>.State: TIMED_WAITING at java.lang.<span class="code-object">Thread</span>.sleep(<span class="code-object">Thread</span>.java:-1) at org.apache.zookeeper.client.StaticHostProvider.next(StaticHostProvider.java:336) at org.apache.zookeeper.ClientCnxn$SendThread.run(ClientCnxn.java:1131)
</pre>
</div></div>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">
<span class="code-quote">"main-EventThread@1594"</span> daemon prio=5 tid=0x10 nid=NA waiting
java.lang.<span class="code-object">Thread</span>.State: WAITING
blocks Curator-ConnectionStateManager-0@1539
blocks main@1
at java.lang.<span class="code-object">Object</span>.wait(<span class="code-object">Object</span>.java:-1)
at java.lang.<span class="code-object">Thread</span>.join(<span class="code-object">Thread</span>.java:1252)
at org.apache.zookeeper.ZooKeeper.testableWaitForShutdown(ZooKeeper.java:2932)
at org.apache.zookeeper.ZooKeeper.close(ZooKeeper.java:1350)
at org.apache.curator.HandleHolder.internalClose(HandleHolder.java:158)
at org.apache.curator.HandleHolder.closeAndReset(HandleHolder.java:84)
at org.apache.curator.ConnectionState.reset(ConnectionState.java:203)
- locked <0x716> (a org.apache.curator.ConnectionState)
at org.apache.curator.ConnectionState.handleExpiredSession(ConnectionState.java:380)
at org.apache.curator.ConnectionState.checkState(ConnectionState.java:315)
at org.apache.curator.ConnectionState.process(ConnectionState.java:169)
at org.apache.zookeeper.ClientCnxn$EventThread.processEvent(ClientCnxn.java:525)
at org.apache.zookeeper.ClientCnxn$EventThread.run(ClientCnxn.java:500)
</pre>
</div></div>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">
<span class="code-quote">"Curator-ConnectionStateManager-0@1539"</span> daemon prio=5 tid=0xe nid=NA waiting <span class="code-keyword">for</span> monitor entry
  java.lang.<span class="code-object">Thread</span>.State: BLOCKED
	 waiting <span class="code-keyword">for</span> main-EventThread@1594 to release lock on <0x716> (a org.apache.curator.ConnectionState)
	  at org.apache.curator.ConnectionState.checkTimeouts(ConnectionState.java:209)
	  at org.apache.curator.ConnectionState.getZooKeeper(ConnectionState.java:96)
	  at org.apache.curator.CuratorZookeeperClient.getZooKeeper(CuratorZookeeperClient.java:164)
	  at org.apache.curator.framework.state.ConnectionStateManager.checkSessionExpiration(ConnectionStateManager.java:307)
	  at org.apache.curator.framework.state.ConnectionStateManager.processEvents(ConnectionStateManager.java:283)
	  - locked <0x7cf> (a org.apache.curator.framework.state.ConnectionStateManager)
	  at org.apache.curator.framework.state.ConnectionStateManager.access$000(ConnectionStateManager.java:44)
	  at org.apache.curator.framework.state.ConnectionStateManager$1.call(ConnectionStateManager.java:120)
	  at java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:266)
	  at java.util.concurrent.FutureTask.run(FutureTask.java:-1)
	  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	  at java.lang.<span class="code-object">Thread</span>.run(<span class="code-object">Thread</span>.java:748)
</pre>
</div></div>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mskonovalov">mskonovalov</a>:</i>
<p>It more looks like a live-lock in sendThread</p>

<p><tt>StaticHostProvider</tt></p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">
<span class="code-keyword">if</span> (needToSleep) {
    <span class="code-keyword">try</span> {
<span class="code-object">Thread</span>.sleep(spinDelay);
    } <span class="code-keyword">catch</span> (InterruptedException e) {
LOG.warn(<span class="code-quote">"Unexpected exception"</span>, e);
    }
}
</pre>
</div></div>
<p>spin delay is 1000</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mskonovalov">mskonovalov</a>:</i>
<p>And honestly I have another issue with this method in ZK in StaticHostProvider.</p>

<p> </p>

<p>We have a high concurrent system and if ZK is not available I want to fail fast. But they have hardcoded 1000ms timeout here so call with 3 retries takes about 4.5 sec even if I pass all the timeouts ~100ms.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>GitHub user Randgalt opened a pull request:</p>

<p>    <a href="https://github.com/apache/curator/pull/283" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/283</a></p>

<p>    <a href="https://issues.apache.org/jira/browse/CURATOR-472" title="Test-only code started to be called in prod" class="issue-link" data-issue-key="CURATOR-472"><del>CURATOR-472</del></a> - fix deadlocks during tests as well as TTL test issues</p>

<p>    Fixed 2 problems: 1) internalClose can be called from a ZooKeeper background thread (as part of a Watcher callback). Calling ZooKeeper.close with a waitForShutdownTimeoutMs in this case will cause a deadlock as no other events can be processed until the current thread exits. 2) All TTL tests must set the system property "zookeeper.extendedTypesEnabled"</p>

<p>You can merge this pull request into a Git repository by running:</p>

<p>    $ git pull <a href="https://github.com/apache/curator" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator</a> <a href="https://issues.apache.org/jira/browse/CURATOR-472" title="Test-only code started to be called in prod" class="issue-link" data-issue-key="CURATOR-472"><del>CURATOR-472</del></a></p>

<p>Alternatively you can review and apply these changes as the patch at:</p>

<p>    <a href="https://github.com/apache/curator/pull/283.patch" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/283.patch</a></p>

<p>To close this pull request, make a commit to your master/trunk branch<br/>
with (at least) the following in the commit message:</p>

<p>    This closes #283</p>

<hr />
<p>commit b91036be07aaddc4ed2d81615b08b33075e923af<br/>
Author: randgalt <randgalt@...><br/>
Date:   2018-12-01T15:55:55Z</p>

<p>    <a href="https://issues.apache.org/jira/browse/CURATOR-472" title="Test-only code started to be called in prod" class="issue-link" data-issue-key="CURATOR-472"><del>CURATOR-472</del></a></p>

<p>    Fixed 2 problems: 1) internalClose can be called from a ZooKeeper background thread (as part of a Watcher callback). Calling ZooKeeper.close with a waitForShutdownTimeoutMs in this case will cause a deadlock as no other events can be processed until the current thread exits. 2) All TTL tests must set the system property "zookeeper.extendedTypesEnabled"</p>

<hr />

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user cammckenzie commented on the issue:</p>

<p>    <a href="https://github.com/apache/curator/pull/283" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/283</a></p>

<p>    Looks good to me. I pushed a fixed to the TestTtlNodes unit test is it was consistently failing for me.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user asfgit closed the pull request at:</p>

<p>    <a href="https://github.com/apache/curator/pull/283" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/283</a></p>
